### ISSUE TITLE
Record based authorization

### ISSUE THREAD

**Issue by danielbecker** [AnalyticsMediaGroup/strongbolt#24]
We would like to use strongbolt in our application due to its ability to authorize on a controller and model level. However, we need to grant/revoke access on a single record basis as well. It would be great if that was supported as well. Is it possible to add that feature in the near future? If you can point us in the right direction we’d be more than happy to help implementing that feature.

**Comment by tobithiel** [comment]
Hmm, we haven't put much thought into that feature. We haven't had the need for it and got along pretty well with the Tenant Model.
That would be quite cumbersome to configure with a lot of models, right? How would that work?

Technically it should be possible to integrate this though. It would mostly be modifying [Strongbolt::UserAbilities](https://github.com/AnalyticsMediaGroup/strongbolt/blob/master/lib/strongbolt/user_abilities.rb). It can't save model specific abilities right now. Afterwards it should be enough to add that behavior to the [populate_capabilities_cache](https://github.com/AnalyticsMediaGroup/strongbolt/blob/master/lib/strongbolt/user_abilities.rb#L95) and [capability_in_cache?](https://github.com/AnalyticsMediaGroup/strongbolt/blob/master/lib/strongbolt/user_abilities.rb#L154) methods. There might be more changes, but these are the key areas, after thinking about this shortly.

**Comment by matissg** [comment]
@tobithiel I'd like to implement your gem in my project and might need to have record level authorization as well. Here is my thinking of how to do it and would be happy if you comment:
1. `Permission` model with appropriate table would be set
2. polymorphic relationship with `User` model would be defined
3. `permitable_type` and `permitable_id` would be Models  (e.g., Post, Article, etc.) which we permit for `current_user`
4. `level` would be defined - `owner`, `read`, `write`, `delete`, etc.

Basically we would give access to certain record via that `Permission` model after **Strongbolt** check general authorization for the target Model. For example, `current_user` wants to access `Post.where(id: 2)`, so **Strongbolt** check if `current_user` can access `Post` model and then through `Permission` model we can find if there is link to `Post with id=2` for `current_user`. If yes, then record is loaded, if no - redirect to errors_path.

**What do you say, would this work with the gem?**

**Comment by tobithiel** [comment]
What's the benefit of having another model `Permission`? It would make more sense in my opinion, to implement it in the existing `Capability` model.

My concern still is, that if you have 2000+ instances of a model, how are you gonna assign them all to the respective users manually in an efficient way. That seems very tedious. I don't see the use case for this right now. Maybe you could explain what you're trying to achieve?

I don't think we're gonna implement this at 605, since we have no use case for that behavior right now. But we'd be happy to accept a PR, if the use case is there.

**Comment by matissg** [comment]
@tobithiel  I think I got your point. I'm thinking I probably can use `user_groups`. If I understand it right, I can place a user in particular group and then set that particular record can be accessed by specific `user_group`. That should work, right?

**Comment by tobithiel** [comment]
Yeah, that should work, if `Capability` can store the record ID or something like that.

**Comment by ojab** [comment]
I'm actually need this as well.
IMO most appropriate place to store this info is `Role` (hey, we're RBAC  after all) and grant access by `role_id`.
So `@results_cache["#{action_model}#{attrs}-#{scope}"]` should be modified to be something like `@results_cache[["#{action_model}#{attrs}", scope]]` where scope can be an `Array`/`Set` of `role_id`.
So we'll check for scope `all`, `action_model.user_id` against scope `current_user.id` and `action_model.role_id` against scope `current_user.roles_ids`.
Sounds like a plan or I'm missing something?

**Comment by ojab** [comment]
There are some other requirements for me and I suppose that it's better to discuss all of this in one place instead of creating multiple issues.

Authorization against multiple models:
```ruby
class Order
  has_one :seller, class_name: 'Company'
  has_one :buyer, class_name: 'Company'
end
class Company
  belongs_to :user # or :role
  has_many :orders
end
```
we need to show `order` to both `seller` and `buyer` and AFAIU right now it's not possible. Most straightforward solution that I've came up with is to allow `authorize_as` to take an `Array`/`Set` of models and grant permission if `user` has the authorization for any of them. In the result we will have something like
```ruby
class Order
  authorize_as: %w[Buyer Seller]
end
```

Authorization via `has_many` association:
```ruby
class Order
  has_many :steps
end
class Step
  belongs_to :order
  belongs_to :user # or :role
end
```
because of different workflows depending on `Order` type we have different sets of `Step` and `Order` should be visible to all `User`s involved. AFAIU right now it's not possible. Again, it can be solved with passing `Array`/`Set` of all involved users, this time in place of [id of the checked model](https://github.com/605data/strongbolt/blob/master/lib/strongbolt/user_abilities.rb#L166).

Thoughts?

**Comment by ojab** [comment]
...and I just understood that `autrhorize_as` takes a model, not relation. So we should do something like `authorize_via %i[buyer seller]` & `authorize_via :steps` accordingly, where `authorize_via` will process relations.

**Comment by matissg** [comment]
@ojab I'm trying to understand your approach regarding record based authorization, however I don't get it. I mean, `Role` allows to set you `name` and then do this parent/child thing. As I understand there has to be some mechanism, how you would store and update info on `user_id` and their rights for particular record. Do you plan somehow to involve `tenant` part of the Strongbolt for that?

**Comment by ojab** [comment]
Record can has `role_id` in addition to (or as replacement for) `user_id`, so we'll check `role_id` against `current_user` roles, that's enough to give specific roles access to the specific records (it solves https://github.com/605data/strongbolt/issues/24#issuecomment-303470313, if I understand requirement correctly).
If the requirement is to gave arbitrary set of users/roles access to arbitrary set of records -- it can be done using `Authorization via has_many association` from https://github.com/605data/strongbolt/issues/24#issuecomment-313874995 and a separate table with permissions (maybe `strongbolt_capabilities` with some modifications can be this table, haven't looked at it thoroughly).

**Comment by ojab** [comment]
Or it can be via `belongs_to` instead of via `has_many` and we just do

**Comment by ojab** [comment]
Or it can be authorization via `belongs_to` instead of `has_many` if we have predefined sets of authorized roles for each record, so we can just do something like
```
class Order
  belongs_to :strongbolt_capability
  authorize_via :strongbolt_capability
end
```

**Comment by ojab** [comment]
(it not necessary will be `belongs_to :strongbolt_capability` relation, since it contain only one action, but the point is: authorization will be done via another table without creating a separate authorization record for each record in `orders` table)

**Comment by ojab** [comment]
..and after actually trying to create roles/user_groups in my toy app, I've realized my mistake and agree that it should actually be `UserGroup`, not 'Role' in all my comments above.

**Comment by tobithiel** [comment]
@ojab Could your situation be addressed by making `ownership` checks more flexible? Or is `ownership` something that you want to consider separate?
Right now ownership requires a method `user_id`, which is compared to the `current_user.id`, but could be extended to make use of ActiveRecord associations.
That's how I understand your `authorize_via` concept, but correct me if I'm wrong.
